### ISSUE TITLE
Incorrect version showed

### ISSUE THREAD

**Issue by SeanFeldman** [paolosalvatori/ServiceBusExplorer#182]
### Steps to repro

- Have an older version running locally (4.0.105).
- Esure to be be connected to the internet
- Version showed will be not the actual version, but the latest available on GitHub (latest release)

I suspect either embedded version is not embedded properly or [`VersionHelper`](https://github.com/paolosalvatori/ServiceBusExplorer/blob/83444ad7e84534e2bd501ca0e00f68b06c84f109/src/ServiceBusExplorer/Helpers/VersionHelper.cs#L52) has a bug.

Originally, there was a [`VersionProvider`](https://github.com/paolosalvatori/ServiceBusExplorer/blob/83444ad7e84534e2bd501ca0e00f68b06c84f109/src/ServiceBusExplorer/Helpers/VersionProvider.cs#L6) to determine product version, but it's no longer used.

[`MainForm` only invokes `VersionHelper`](https://github.com/paolosalvatori/ServiceBusExplorer/blob/067bdc000a121e62ed808bf1c7e02b265f7b5171/src/ServiceBusExplorer/Forms/MainForm.cs#L7003), which seems to be the issue.

**Comment by SeanFeldman** [comment]
@paolosalvatori do you recall why the code was switched to retreive version from GH rather that looking at embedded version?

**Comment by paolosalvatori** [comment]
Yes. Because when creating a release, the version was not visible in the titlebar of the tool and it was invariantly equal to 1.0.0

**Comment by SeanFeldman** [comment]
> when creating a release

Are you referring to a released version that could be downloaded from GitHub (https://github.com/paolosalvatori/ServiceBusExplorer/releases/latest)?

**Comment by paolosalvatori** [comment]
Yes :)

**Comment by SeanFeldman** [comment]
In that case, the version should be baked **into** the released version. If it's not, then it's a bug. We shouldn't get a version from GH as what you run is not always the latest and incorrectly will report as the latest. Would you agree @paolosalvatori ?

If yes, I'll send a PR. Otherwise, please elaborate why you disagree. Thanks.

**Comment by SeanFeldman** [comment]
Here's what's happening. I'm running a 4.0.106 on my machine. Do to #187, a new release was created, but I've never updated to that and still, locally the version is reported as 4.0.107.

![image](https://user-images.githubusercontent.com/1309622/36126899-efa366ea-1017-11e8-825a-57c3fd4e6348.png)

Looking at the file properties, and what's stamp GitVersion has applied on the build server when the bits were packaged, it looks right.

![image](https://user-images.githubusercontent.com/1309622/36126933-157ea9b0-1018-11e8-8fc1-20550f2ae579.png)

And that's why I think retrieving version from GH is wrong @paolosalvatori 🙂

**Comment by paolosalvatori** [comment]
Yes... and I'm aware of this problem. This is just a workaround to avoid to show 1.0.0 any time you start the exe coming from any release.

**Comment by SeanFeldman** [comment]
> This is just a workaround to avoid to show 1.0.0 any time you start the exe coming from any release.

Could you please elaborate on what do you mean by "any time you start the exe coming from any release."? Thank you.

**Comment by paolosalvatori** [comment]
It means that if I don't make a call to retrieve the version number of the latest release from GitHub, 1.0.0 would be shown in the title bar of the main window :) If you remember, we had this conversation last year or so. :)

**Comment by SeanFeldman** [comment]
@paolosalvatori don't think that was the case. The issue I remember is that[ locally built bits didn't have the version other than 1.0.0](https://github.com/paolosalvatori/ServiceBusExplorer/issues/125#issuecomment-310123019). Which, BTW, was [intentional](https://github.com/paolosalvatori/ServiceBusExplorer/pull/104). After that the code was modified to read from GitHub all the time.

Suggest we stick with the following:
- Released versions on GH to have the build number driven by the repo (GitVersion).
- Locally compited version to always carry version 1.0.0 to identify it's a local build and not an officially released one.

**Comment by paolosalvatori** [comment]
@SeanFeldman you are right, that was the issue. By the way, you decided that this behavior was intentional, not me. :) I told you that I'd prefer locally built copies to have the right version, and not 1.0.0. :)

**Comment by SeanFeldman** [comment]
Currently, the code always [looks up GitHub](https://github.com/paolosalvatori/ServiceBusExplorer/blob/c84dc37623f9c8134f1d9554f52ec1442b20afee/src/ServiceBusExplorer/Forms/MainForm.cs#L7003) for the latest version, which is incorrect. Event if you are developing locally on an outdated version. 

Could you please elaborate on the need to have "locally built copies to have the right version, and not 1.0.0" @paolosalvatori ? What is the problem that you find with having the locally built version being 1.0.0?

**Comment by paolosalvatori** [comment]
@SeanFeldman I know, and if you remember I told you that I didn't like this solution as it always shows the latest version, even if the current version of the tool is older. A lot of people download the code and compile it using Visual Studio. If we show 1.0.0 it's the same as showing the latest version available on GitHub: it's fundamentally wrong and I can't understand what is the real version of the tool. The version should be taken from AssemblyInfo.cs not from GitHub. But you need to provide a way to update the AssemblyInfo any time you create a new release of the tool. This way, both the compiled version and the release would show the correct version number. Makes sense? Please, do not reply with a question, I think my assert is clear enough this time. :)

**Comment by SeanFeldman** [comment]
No more questions. If you want it this way, be it so. I'll outline what's wrong with the curent state and you can decide how to address it.

- When someone uses the tool as-is, they most likely would get it from GitHub download page or Chocolatey. The version presented is never going to be correct if a newer version of the tool was released.
- When someone downloads the code and compiles it locally, and later reports there's a problem, it's not possible to associate it with the correct branch.

**Comment by paolosalvatori** [comment]
@SeanFeldman let me clarify. I'd like to eliminate the current implementation and use the following approach:

-

**Comment by tomkerkhove** [comment]
But does that really make sense? Shouldn't we keep an automated process for this instead where this is done for us and we can release whenever we want, without having to do source code changes?

If they want to download the sources, they can still go to the release which has the source code linked to it? https://github.com/paolosalvatori/ServiceBusExplorer/releases/tag/4.0.107

Or check the git tags?

**Comment by danielmarbach** [comment]
@paolosalvatori I'm not sure I can entirely follow your reasoning here. One of the benefits of using an automated approach as outlined by @SeanFeldman is that it can be made sure the version is always up to date and correctly incremented. Furthermore, when someone looks at the UI it is clear how the version relates to commits and can always be reproduced. So why would you want to go back to a manual approach? 

I would agree that having a version of 1.0.0 for local builds seems weird. But that could be mitigated by using GitVersion. GitVersion can always reproduce a sane version based on the locally checked out git history. For example, if I'd pull down explorer and create a feature branch that feature branch would be locally versioned properly.

**Comment by SeanFeldman** [comment]
What @danielmarbach makes sense. If version needs to be generated locally, it could leverage GitVersion local info. Note: AppVeyor build number will be missing (currently used as a `build number` in the `major.minor.build` sceme`).

A few examples:

### Information available locally 
#### `develop` branch
```json
{
  "Major":4,
  "Minor":1,
  "Patch":0,
  "PreReleaseTag":"unstable.182",
  "PreReleaseTagWithDash":"-unstable.182",
  "PreReleaseLabel":"unstable",
  "PreReleaseNumber":182,
  "BuildMetaData":"",
  "BuildMetaDataPadded":"",
  "FullBuildMetaData":"Branch.develop.Sha.debb75136d907404e67c0382840a655df44c3cf0",
  "MajorMinorPatch":"4.1.0",
  "SemVer":"4.1.0-unstable.182",
  "LegacySemVer":"4.1.0-unstable182",
  "LegacySemVerPadded":"4.1.0-unstable0182",
  "AssemblySemVer":"4.1.0.0",
  "FullSemVer":"4.1.0-unstable.182",
  "InformationalVersion":"4.1.0-unstable.182+Branch.develop.Sha.debb75136d907404e67c0382840a655df44c3cf0",
  "BranchName":"develop",
  "Sha":"debb75136d907404e67c0382840a655df44c3cf0",
  "NuGetVersionV2":"4.1.0-unstable0182",
  "NuGetVersion":"4.1.0-unstable0182",
  "CommitsSinceVersionSource":182,
  "CommitsSinceVersionSourcePadded":"0182",
  "CommitDate":"2018-02-15"
}
```
#### master branch

```json
{
  "Major":4,
  "Minor":0,
  "Patch":107,
  "PreReleaseTag":"",
  "PreReleaseTagWithDash":"",
  "PreReleaseLabel":"",
  "PreReleaseNumber":"",
  "BuildMetaData":"",
  "BuildMetaDataPadded":"",
  "FullBuildMetaData":"Branch.master.Sha.c84dc37623f9c8134f1d9554f52ec1442b20afee",
  "MajorMinorPatch":"4.0.107",
  "SemVer":"4.0.107",
  "LegacySemVer":"4.0.107",
  "LegacySemVerPadded":"4.0.107",
  "AssemblySemVer":"4.0.107.0",
  "FullSemVer":"4.0.107",
  "InformationalVersion":"4.0.107+Branch.master.Sha.c84dc37623f9c8134f1d9554f52ec1442b20afee",
  "BranchName":"master",
  "Sha":"c84dc37623f9c8134f1d9554f52ec1442b20afee",
  "NuGetVersionV2":"4.0.107",
  "NuGetVersion":"4.0.107",
  "CommitsSinceVersionSource":0,
  "CommitsSinceVersionSourcePadded":"0000",
  "CommitDate":"2018-02-12"
}
```

#### custom branch
```json
{
  "Major":4,
  "Minor":0,
  "Patch":105,
  "PreReleaseTag":"csharp-language-version.47",
  "PreReleaseTagWithDash":"-csharp-language-version.47",
  "PreReleaseLabel":"csharp-language-version",
  "PreReleaseNumber":47,
  "BuildMetaData":"",
  "BuildMetaDataPadded":"",
  "FullBuildMetaData":"Branch.csharp-language-version.Sha.174094283b95c945e7962ce1162af69e18695865",
  "MajorMinorPatch":"4.0.105",
  "SemVer":"4.0.105-csharp-language-version.47",
  "LegacySemVer":"4.0.105-csharp-language-ve47",
  "LegacySemVerPadded":"4.0.105-csharp-language-0047",
  "AssemblySemVer":"4.0.105.0",
  "FullSemVer":"4.0.105-csharp-language-version.47",
  "InformationalVersion":"4.0.105-csharp-language-version.47+Branch.csharp-language-version.Sha.174094283b95c945e7962ce1162af69e18695865",
  "BranchName":"csharp-language-version",
  "Sha":"174094283b95c945e7962ce1162af69e18695865",
  "NuGetVersionV2":"4.0.105-csharp-language-0047",
  "NuGetVersion":"4.0.105-csharp-language-0047",
  "CommitsSinceVersionSource":47,
  "CommitsSinceVersionSourcePadded":"0047",
  "CommitDate":"2018-02-15"
}
```

My only concern would be how to differentiate a properly released build number from a locally compiled one. For example, someone pulls down master, modifies locally, compiles and has a version `4.0.107` which is a "duplicate" of an officially released `4.0.107` via AppVeyor which corresponds to a different set of commits/changes.

**Comment by paolosalvatori** [comment]
@SeanFeldman @tomkerkhove @danielmarbach I don't want to go back to a manual approach, actually, I'm saying the opposite thing. Is there a way to add the version number to the AssemblyInfo.cs or any other file whenever we release a new version so that the tool can read the current version from a known place also when someone compiles the code locally? :)

**Comment by SeanFeldman** [comment]
@paolosalvatori see https://github.com/paolosalvatori/ServiceBusExplorer/issues/182#issuecomment-367057511

**Comment by SeanFeldman** [comment]
This is why the current GH lookup is making things awkward.
I've added a feature and running locally, the app says `4.0.108` which is nonsense as this feature was never released...

![image](https://user-images.githubusercontent.com/1309622/37868545-fc87cce6-2f6d-11e8-968b-639061dfa141.png)

**Comment by ErikMogensen** [comment]
As mentioned previously in this thread, GitVersion, may be a solution for this. What do you think? I do not have any experience with it, but I could do some testing. It has support for Appveyor, https://gitversion.readthedocs.io/en/latest/build-server-support/build-server/appveyor/.

**Comment by ErikMogensen** [comment]
Is is possible to add GitVersion support to the build from inside Visual Studio and still be able to set version when building with AppVeyor? Preferably there should be some metadata that says what kind of build it is.

**Comment by SeanFeldman** [comment]
GitVersion doesn't care about VS. It looks at Git information such as branch name, commit sha, etc.
So building locally or on CI server would be consistent.

**Comment by ErikMogensen** [comment]
What about using the GitVersion msbuild task? Then we have a version on both the VS and the appveyor builds. Preferably we should be able to tell the difference between the builds, but since we do not offer official support for neither build it is not a big deal. The highest priority now must be to release a new version with a built in version number instead of retrieving it from GitHub, since the longer we wait the more SBE instances there will be using that way of versioning.

**Comment by SeanFeldman** [comment]
> The highest priority now must be to release a new version with a built in version number instead of retrieving it from GitHub, since the longer we wait the more SBE instances there will be using that way of versioning.

Preaching to the choir. I've failed to persuade @paolosalvatori that that's what needs to happen. Perhaps he has changed his mind since then.

GitVersion MSBuild task what I mean by "GitVersion". It plugs into MSBuild pipeline, which is everywhere, locally or build server.

**Comment by ErikMogensen** [comment]
I see. As I understand it the problem with GitVersion is that it won't a handle a plain download from GitHub. I know that some organizations don't allow git communication with GitHub so plain downloads is the only way to get SBE with source code for many people. 

I thought these versioning requirements would be a common problem but I could not find any good solutions on the internet. To get something working short term I suggest we switch to manually setting assembly file version when we make a new release. It is not ideal but better than the current solution.

**Comment by ErikMogensen** [comment]
> Plain download of what? In case artifacts are downloaded from releases pages or the latest version page, those are generated on the build server and will be versioned. 

When downloading the circled item below from https://github.com/paolosalvatori/ServiceBusExplorer/releases you get the source code without any version information emdedded into it. 
![image](https://user-images.githubusercontent.com/9644248/48442354-4e4f6b00-e78e-11e8-8bac-6c5a16a50450.png)

> Anyone who wants to pull down and compile their own code would have it as 1.0.0 OOTB or can set their own version 

I also think that you should get the source code with a version number if you pull it down. I believe that was the reason the functionality for getting the version information from GitHub was implemented.

**Comment by SeanFeldman** [comment]
The assets you're point to are functionality of GitHub whenever a new release is created.
Are you expecting any repository on GitHub to have a version embedded in the downloadable source attributed to a certain release? I'd say that's unrealistic expectation and unnecessary. 

Look at the ASB client. It has versioned releases and no version is embedded in the **source** code zip file. For that reason, I'm going to object adding info where it doesn't belong. If someone wants a source w/o git info, they can download this zip. If someone wants a version, they have choco or **release** zip (top item in the list you referenced). Someone who wants to customize and run their own build, that's their business.

**Comment by danielmarbach** [comment]
Agree with @seanfeldman

**Comment by SeanFeldman** [comment]
> It has version information in the zipped source code

Exactly my point! 😃 That's because they are **manually** manage version and check it into the source code. It's _their choice_ to do it that way. And if one day they release incorrect version, it will be their responsibility. What we've done is we said "master is what's released". When a `release` branch is merged into `master` with all the changes (bug fixes, improvements, features), version is automatically updated and no human error can take place. Whatever is compiled locally - not maintainers responsibility. I do not see why it has to be...

**Comment by paolosalvatori** [comment]
> > The highest priority now must be to release a new version with a built in version number instead of retrieving it from GitHub, since the longer we wait the more SBE instances there will be using that way of versioning.
> 
> Preaching to the choir. I've failed to persuade @paolosalvatori that that's what needs to happen. Perhaps he has changed his mind since then.
> 
> GitVersion MSBuild task what I mean by "GitVersion". It plugs into MSBuild pipeline, which is everywhere, locally or build server.

Sorry guys for coming late to this conversation, but it was tough week. My point is and has always been that the tool should show the same version no matter if you build it in Visual Studio or build it with AppVeryor. The reason why I included some code to call the GitHub API to retrieve from the version of the most recent release (which could be wrong version by the way) is that I don't like to see 1.0.0 when I build the tool locally.

**Comment by ErikMogensen** [comment]
I will summarize how I perceive it, please correct me if I am wrong:

The GitVersion MSBuild Task, which is in use right now, creates a version number on the fly when building using only information in the GitHub history. This version number is embedded in the assembly and nowhere else. This works for both AppVeyor and Visual Studio builds. 

I want the version number to be persisted in the source code, otherwise you'll not have any version number at all if you download the just the source code. This applies to source code downloads from https://github.com/paolosalvatori/ServiceBusExplorer/releases and when choosing _Download Zip_ from https://github.com/paolosalvatori/ServiceBusExplorer. 

It looks like it is difficult to create a version number automatically when building. My suggestion is to do like they do with the ASB client, https://github.com/Azure/azure-service-bus-dotnet, set a version number manually in the source code until we have found a solution that automatically increments version.

**Comment by SeanFeldman** [comment]
@paolosalvatori 

> My point is and has always been that the tool should show the same version no matter if you build it in Visual Studio or build it with AppVeryor. 

Yep, that's what you get when the repo is _cloned_.  When source code is copied, that will not happen.

> The reason why I included some code to call the GitHub API to retrieve from the version of the most recent release (which could be wrong version by the way) is that I don't like to see 1.0.0 when I build the tool locally.

When building locally w/o git info, I would argue that's desirable to distinguish what's being built _properly_ vs something that was not. If a binary is floating in a company, how do you know it's an authentic version or something butchered and compiled on someone's machine, later published on a network share for others to use?

@ErikMogensen 

> The GitVersion MSBuild Task, which is in use right now, creates a version number on the fly when building using only information in the GitHub history

Using **Git** information. Therefore if a repo is cloned, building locally would have the version and branch name. Using MSBuild, VS, or build-server.

> I want the version number to be persisted in the source code, otherwise you'll not have any version number at all if you download the just the source code.

Are you downloading sources @ErikMogensen or are you cloning the repo? The days of getting a zipped file for continuous work are kind of over. If you're serious about your intentions, you'd want the repo, not the zip with sources.

> It looks like it is difficult to create a version number automatically when building.

I'm not sure why you've came to this conclusion. We've been versioning for quite a while.

> My suggestion is to do like they do with the ASB client, https://github.com/Azure/azure-service-bus-dotnet, set a version number manually in the source code until we have found a solution that automatically increments the version number. That will create a risk that the embedded version number does not match the version in GitHub.

Why would you want to go with the manual route which has a serious flaw and doesn't really bring any value is beyond my comprehension. The ASB team is doing versioning this way because it's better. They do it this way because they have way too much work to fix how they do it. And a lot of the things they've done with that repo so far I've had some involvement with and insights.

I also have a couple of small projects that I maintain where I tried the manual route as well. Huge drawback to that is manual steps and remembering to do it right.

**Summary**: at this point, I'm out of energy to try to convince not to invest into a bad idea. Automation is for a reason. If you wish to follow a manual process, fine. I'm not going to support it, but I'm not going to stand in your way either.

**Comment by ErikMogensen** [comment]
First of all, sorry for draining your energy.

> > It looks like it is difficult to create a version number automatically when building.
> I'm not sure why you've came to this conclusion. We've been versioning for quite a while.

What I should have written was: It looks like it is difficult to update a version number which is embedded in the source code automatically when building.

> I'm not going to support it, but I'm not going to stand in your way either.

Same goes for me whatever the outcome of this as long as the fetching of version number from GitHub is removed.

**Comment by paolosalvatori** [comment]
@ErikMogensen @SeanFeldman if this is an issue, let's remove the call to retrieve the version of the latest release on GitHub, and that's it. I'm definitely not happy to see 1.0.0. In this case, I will remove the version from the title bar and I'll leave it in the About form.

**Comment by paolosalvatori** [comment]
@ErikMogensen if it's a matter of reducing the delay with which SBE starts, I can surely change the call so that if it doesn't return in maximum 500 ms, the call to GitHub aborts. :)

**Comment by ErikMogensen** [comment]
Ah! :)  Then it is something else, my next guess is it is the downloading of debug symbols from Microsoft. That is a of course a different matter.

**Comment by SeanFeldman** [comment]
Have a look at [MinVer](https://www.nuget.org/packages/MinVer/1.0.0-beta.1). As simple as adding a package reference and version can be driven by the Git tags. Will generate correct version locally when code is **cloned**, which should be always the case.

**Comment by danielmarbach** [comment]
It makes sense to switch to SDK style proj anyway since this is the new prefered format which comes with so much benefits like cleanliness, nuget packaging automation and more

**Comment by SeanFeldman** [comment]
👍 to what @danielmarbach said.
There's an issue for that, #116, which I've reopened and can be tackled first to unblock this one should MinVer be used.

**Comment by ErikMogensen** [comment]
I can remove the version information from the title. 

When building SBE without having downloaded it the Git way it will display 0.0.0.0 as version number in the About box. I think it is better to display 0.0.0.0 than getting the latest version number from GitHub, since that may not be that version executing and then the About form would be lying which would be confusing.
